sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java

```diff
@@ -0,0 +1,109 @@
+     */
+    private final Output<Map<String, Output<?>>> outputs;
+
+    /**
+     * Create a Stack with stack outputs defined in derived class constructor.
+     *
+     * @param projectName             the current project name
+     * @param stackName               the current stack name
+     * @param resourceTransformations optional list of transformations to apply to this stack's resources during construction.
+     *                                The transformations are applied in order, and are applied after all the transformations of custom
+     *                                and component resources in the stack.
+     * @param outputs                 the stack outputs
+     */
+    private Stack(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations,
+            Output<Map<String, Output<?>>> outputs
+    ) {
+        super(
+                RootPulumiStackTypeName,
+                rootPulumiStackName(projectName, stackName),
+                ComponentResourceOptions.builder().resourceTransformations(
+                        resourceTransformations
+                ).build()
+        );
+        this.outputs = requireNonNull(outputs);
+    }
+
+    private static String rootPulumiStackName(String projectName, String stackName) {
+        return String.format("%s-%s", requireNonNull(projectName), requireNonNull(stackName));
+    }
+
+    @InternalUse
+    public Output<Map<String, Output<?>>> getOutputs() {
+        return this.outputs;
+    }
+
+    @InternalUse
+    public static Function<Supplier<Map<String, Output<?>>>, Stack> factory(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations
+    ) {
+        // Stack initialization requires specific sequence of events:
+        // - create a Stack instance (note that the super constructor calls readOrRegisterResource)
+        // - set the Stack instance in the global state (user code will try to reach it via Resource constructor)
+        // - register the Stack outputs in the Pulumi engine
+        // - call the user code callback (asynchronously to avoid any additional cyclic dependency problems)
+        // - complete the Stack initialization (by completing the lazyFuture)
+        return (Supplier<Map<String, Output<?>>> callback) -> {
+            // incomplete future that will be completed after the Stack is constructed
+            var lazyFuture = new CompletableFuture<Map<String, Output<?>>>();
+            // incomplete outputs that will be completed when lazyFuture completes
+            var lazyOutputs = Output.of(lazyFuture);
+            // Stack must be the first Resource to be constructed,
+            // before any user code attempts to construct any Resource
+            var stack = new Stack(projectName, stackName, resourceTransformations, lazyOutputs);
```

Comment:
I have some doubts this sequencing works, I worry about a race. Does `new Stack()` launch `RegisterResource` or some other async call in a bg thread? Or it doesn't because it's a Component? Need to read it carefully. 

It seems that it *does* call RegisterResource to the ResourceMonitor gRPC service.

Now that will race with `deploymentInstance.setStack(stack)` and  `DeploymentInternal.getInstance().registerResourceOutputs(stack, lazyOutputs);`. This may or may not be okay? Would something call `deploymentInstance.getStack()` before the RegisterResource thing?

Reply:
Great points. My understanding is that `getStack()` will only be called for all other resources then the `Stack resource.

Currently there are two places where `getStack` is called:
- (read or register resource) `prepareResourceAsync` -> `getRootResourceAsync` - if a resource has no parent and is not a Stack
- Resource constructor - if a resource has no parent and is not a Stack

Reply:
Regarding outputs registration, AFAIU, `lazyOutputs` being not completed will block the registration until completed via `whenComplete`.

Reply:
Good! In that case we should be OK here!